### PR TITLE
protecting entities using ACLs

### PR DESCRIPTION
**Pull request by xabbuh**

### CODE REVIEW

File: cookbook/entities/acl.rst [hunk 1]
```diff
@@ -0,0 +1,246 @@
...
+    {
+        // ...
+    }
+
+After you have done this and have cleared the cache you can toggle all kinds of permission checks
+(``CREATE``, ``EDIT``, ``DELETE``, ``VIEW``, and ``ASSIGN`` in the user role management interface.
+
+.. tip::
+
+    You can use the optional group attributes to create visual groups in the UI that can help the
```
Comment (yurio) [inline cookbook/entities/acl.rst]: Now group have another behavior (from https://github.com/orocrm/platform/blob/master/src/Oro/Bundle/SecurityBundle/Resources/doc/implementation.md):

`group_name` parameter is used to group entities by applications. It is used to split security into application scopes.

File: cookbook/entities/acl.rst [hunk 2]
```diff
@@ -0,0 +1,246 @@
...
+    user find your entity.
+
+.. _coobook-entities-acl-create:
+
+Creating Access Control Lists
+-----------------------------
+
+You have to options to define your custom access control lists:
```
Comment (yurio) [inline cookbook/entities/acl.rst]: Please change `to` to `two`

File: cookbook/entities/acl.rst [hunk 3]
```diff
@@ -0,0 +1,246 @@
...
+               // ...
+           }
+       }
+
+   Using the ``@Acl`` annotation does not only create new access control lists to which you can
+   refer in other parts of your code it will also trigger the access decision manager when your
+   actions are accessed by users and thus protect them from being accessed without the needed
+   permissions.
+
+#. If you do not want to protect any controller methods or if you prefer to keep the definition of
```
Comment (yurio) [inline cookbook/entities/acl.rst]: Please add description that this definition have parameters `bindings` and `group_name`.
`group_name` the same is in Acl annotation. 
With `bindings` developer can specify controller/action where this definition will work. For example:

`
test_controller:
  type: entity
  class: AcmeBundle\Entity\SomeEntity
  permission: VIEW
  group_name: Test Group
  label: Test controller
  bindings:
    - { class: Acme\DemoBundle\Controller\TestController, method: testAction }
`

File: cookbook/entities/acl.rst [hunk 4]
```diff
@@ -0,0 +1,246 @@
...
+       use AppBunde\Entity\Task;
+       use Oro\Bundle\SecurityBundle\Annotation\Acl;
+       use Symfony\Bundle\FrameworkBundle\Controller\Controller;
+       use Symfony\Component\HttpFoundation\Request;
+
+       class TaskController extends Controller
+       {
+           /**
+            * @Acl(
+            *   id="app_task_view",
```
Comment (yurio) [inline cookbook/entities/acl.rst]: Please add description what id parameter for.

File: cookbook/entities/acl.rst [hunk 5]
```diff
@@ -0,0 +1,246 @@
...
+        app-tasks-grid:
+            # ...
+            actions:
+                # ...
+                edit:
+                    type: navigate
+                    label: Edit
+                    link: update_link
+                    icon: edit
+                    acl_resource. app_task_edit
```
Comment (yurio) [inline cookbook/entities/acl.rst]: Please change to `acl_resource: app_task_edit`

File: cookbook/index.rst [hunk 6]
```diff
@@ -9,6 +9,9 @@ This book contains list of most regularly used solutions in OroPlatform.
   * :doc:`/cookbook/entities/creating_entities`
   * :doc:`/cookbook/entities/data_fixtures`
   * :doc:`/cookbook/entities/datagrid`
+  * :doc:`/cookbook/entities/managing_tasks`
+  * :doc:`/cookbook/entities/deleting_tasks`
```
Comment (yurio) [inline cookbook/index.rst]: Previous links was about all entities, and this two links describes only tasks. I think it will be better to name they just `managing_entities` and `deleting_entities`. And this files should be start with `For example, let's take a task entity...`

File: cookbook/entities/deleting_tasks.rst [hunk 7]
```diff
@@ -0,0 +1,113 @@
+How to Delete Tasks
```
Comment (yurio) [inline cookbook/entities/deleting_tasks.rst]: This file describes how to configure deletion only for the grid. So, i think you should specify this at the header.

File: cookbook/entities/acl.rst [hunk 8]
```diff
@@ -0,0 +1,246 @@
...
+#. :ref:`Activate ACL checks for your entities <coobook-entities-acl-enable>`.
+
+#. :ref:`Create access control lists for all available actions <coobook-entities-acl-create>`.
+
+#. :ref:`Add access checks<coobook-entities-acl-check>` to where your entities are displayed or
+   manipulated.
+
+.. _coobook-entities-acl-enable:
+
+Activating ACL Checks on your Entities
```
Comment (yurio) [inline cookbook/entities/acl.rst]: You have no say anything about ownership type configuration and what developer can limit collection of available permissions for entity.

File: cookbook/entities/acl.rst [hunk 9]
```diff
@@ -0,0 +1,246 @@
...
+       use Oro\Bundle\SecurityBundle\Annotation\Acl;
+       use Symfony\Bundle\FrameworkBundle\Controller\Controller;
+       use Symfony\Component\HttpFoundation\Request;
+
+       class TaskController extends Controller
+       {
+           /**
+            * @Acl(
+            *   id="app_task_view",
+            *   type="entity",
```
Comment (yurio) [inline cookbook/entities/acl.rst]: Please add the description that `type` parameter could be `action` and the difference between them

File: cookbook/entities/acl.rst [hunk 10]
```diff
@@ -0,0 +1,246 @@
...
+           type: entity
+           class: AppBunde\Entity\Task
+           permission: VIEW
+
+.. _coobook-entities-acl-check:
+
+Perform Access Checks
+---------------------
+
+Once you have configured the ACLs you can protect all parts of your application:
```
Comment (yurio) [inline cookbook/entities/acl.rst]: Please add the description that developer can check access via securiy facade service in the code.

File: cookbook/entities/deleting_tasks.rst [hunk 11]
```diff
@@ -0,0 +1,113 @@
...
+#. :ref:`Create a controller that deletes a task <cookbook-entity-delete-controller>`.
+
+#. :ref:`Link to the controller from the data grid <cookbook-entity-delete-grid-config>`.
+
+.. _cookbook-entity-api-manager-service:
+
+Configure the API Manager Service
+---------------------------------
+
+Fortunately, the OroSoapBundle comes with the ``oro_soap.manager.entity_manager.abstract`` service
```
Comment (yurio) [inline cookbook/entities/deleting_tasks.rst]: Please change this sentence. Not only this service can delete Doctrine's entities.

File: cookbook/entities/deleting_tasks.rst [hunk 12]
```diff
@@ -0,0 +1,113 @@
...
+        app.task_manager.api:
+            class: Oro\Bundle\SoapBundle\Entity\Manager\ApiEntityManager
+            parent: oro_soap.manager.entity_manager.abstract
+            arguments:
+                - AppBundle\Entity\Task
+                - '@doctrine.orm.entity_manager'
+
+.. _cookbook-entity-delete-controller:
+
+Create the Controller
```
Comment (yurio) [inline cookbook/entities/deleting_tasks.rst]: Please add configuration of routing.yml file for this controller.

File: cookbook/entities/deleting_tasks.rst [hunk 13]
```diff
@@ -0,0 +1,113 @@
...
+    class TaskController extends RestController
+    {
+        public function deleteAction($id)
+        {
+            return $this->handleDeleteRequest($id);
+        }
+
+        public function getForm()
+        {
+            // this method is not needed to delete entities
```
Comment (yurio) [inline cookbook/entities/deleting_tasks.rst]: Please add description, than now we create only delete API. So, that is why this methods are empty. If developer want to create full API, this methods should contain the code.

File: cookbook/entities/deleting_tasks.rst (hunk 12 above)
Comment (yurio) [inline cookbook/entities/deleting_tasks.rst]: Please specify what we will do delete via API DELETE request.

File: cookbook/entities/deleting_tasks.rst [hunk 14]
```diff
@@ -0,0 +1,113 @@
+How to Delete Tasks
+===================
+
+The data grid that you created so far is able to let the user delete entries without having to
```
Comment (yurio) [inline cookbook/entities/deleting_tasks.rst]: After reading this sentence, i have decide that platform allows to delete the records with reloading the page. Is this true?

File: cookbook/entities/deleting_tasks.rst [hunk 15]
```diff
@@ -0,0 +1,113 @@
...
+
+Update the Data Grid Configuration
+----------------------------------
+
+In your data grid configuration you need to modify two keys:
+
+* Use the ``properties`` key to define the URL that a ``DELETE`` request should be sent to when a
+  task should be removed.
+
+* Add an icon to each row which triggers the actual request when the user clicks on that icon.
```
Comment (yurio) [inline cookbook/entities/deleting_tasks.rst]: Whow should i add this icons to `each row`?

File: cookbook/entities/managing_tasks.rst [hunk 16]
```diff
@@ -0,0 +1,254 @@
...
+        }
+
+        public function configureOptions(OptionsResolver $resolver)
+        {
+            $resolver->setDefaults(array(
+                'data_class' => 'AppBundle\Entity\Task',
+            ));
+        }
+    }
+
```
Comment (yurio) [inline cookbook/entities/managing_tasks.rst]: Please add `More info about Symfony forms you can find here http://symfony.com/doc/current/book/forms.html`

File: cookbook/entities/deleting_tasks.rst [hunk 17]
```diff
@@ -0,0 +1,113 @@
...
+                id: ~
+                delete_link:
+                    type: url
+                    route: app_api_delete_task
+                    params:
+                        - id
+                # ...
+            actions:
+                # ...
+                delete:
```
Comment (yurio) [inline cookbook/entities/deleting_tasks.rst]: Please add the note that at this example, all users will be able to delete all the records from the grid and the link to acl.rst file where user can read how to restrict access

File: cookbook/entities/managing_tasks.rst [hunk 18]
```diff
@@ -0,0 +1,254 @@
...
+.. _cookbook-entity-controller:
+
+The Controllers
```
Comment (yurio) [inline cookbook/entities/managing_tasks.rst]: Please add routing.yml file config for this conntroller

File: cookbook/entities/managing_tasks.rst [hunk 19]
```diff
@@ -0,0 +1,254 @@
...
+        }
+    }
+
+.. _cookbook-entity-template:
+
+The Template
+------------
+
+The template that is responsible to display the form fields should extend the base template
+``OroUIBundle:actions:html.twig`` from the OroUIBundle. This templates defines some basic blocks
```
Comment (yurio) [inline cookbook/entities/managing_tasks.rst]: The template should be extended from OroUIBundle:actions:update.html.twig

File: cookbook/entities/acl.rst (hunk 10 above)
Comment (yurio) [inline cookbook/entities/acl.rst]: Please add the description how to check access in twig templates.

File: cookbook/entities/managing_tasks.rst [hunk 20]
```diff
@@ -0,0 +1,254 @@
...
+        {# the cancelButton() macro creates a button that discards the
+           entered data and leads the user to the linked controller #}
+        {{ UI.cancelButton(path('app_task_index')) }}
+
+        {# the dropdownSaveButton() macro offers a way to let the user select
+           between different options when saving an entity, the selected option
+           will be passed to the controller handling the request as an additonal
+           parameter #}
+        {{ UI.dropdownSaveButton({
+            'html': UI.saveAndCloseButton() ~ UI.saveAndStayButton()
```
Comment (yurio) [inline cookbook/entities/managing_tasks.rst]: There is a case when after creating the entity, user will not be able to edit the created record. We should check this case, so the right code will be:

```
{% set html = UI.saveAndCloseButton() %}
{% if task.id or resource_granted('app_task_update') %}
     {% set html = html ~ UI.saveAndStayButton() %}
 {% endif %}
 {{ UI.dropdownSaveButton({'html': html}) }}
```

File: cookbook/entities/managing_tasks.rst [hunk 21]
```diff
@@ -0,0 +1,254 @@
...
+        }) }}
+    {% endblock navButtons %}
+
+    {% block pageHeader %}
+        {% if task.id %}
+            {% set breadcrumbs = {
+                'entity': task,
```
Comment (yurio) [inline cookbook/entities/managing_tasks.rst]: Please describe this parameters

File: cookbook/entities/managing_tasks.rst [hunk 22]
```diff
@@ -0,0 +1,254 @@
...
+        {% else %}
+            {% set title = 'oro.ui.create_entity'|trans({ '%entityName%': 'Task' }) %}
+            {{ include('OroUIBundle::page_title_block.html.twig', { title: title }) %}
+        {% endif %}
+    {% endblock pageHeader %}
+
+    {% block content_data %}
+        {% set id = 'task-edit' %}
+        {% set dataBlocks = [{
+                'title': 'General'|trans,
```
Comment (yurio) [inline cookbook/entities/managing_tasks.rst]: Please describe this parameters

File: cookbook/entities/managing_tasks.rst [hunk 23]
```diff
@@ -0,0 +1,254 @@
...
+                    'data': [
+                        form_row(form.subject),
+                        form_row(form.description),
+                        form_row(form.dueDate),
+                        form_row(form.priority),
+                    ]
+                }]
+            }]
+        %}
+
```
Comment (yurio) [inline cookbook/entities/managing_tasks.rst]: The entity can be extendable. So it can contain custom fields. In this case we should add this custom fields with code

 ```
   {% set additionalData = [] %}
    {% for child in form.children if child.vars.extra_field is defined and child.vars.extra_field %}
        {% set additionalData = additionalData|merge([form_row(child)]) %}
    {% endfor %}
    {% if additionalData is not empty %}
        {% set dataBlocks = dataBlocks|merge([{
            'title': 'Additional'|trans,
            'subblocks': [{
                'title': '',
                'useSpan': false,
                'data' : additionalData
            }]
        }] ) %}
    {% endif %}
```

File: cookbook/entities/managing_tasks.rst [hunk 24]
```diff
@@ -0,0 +1,254 @@
...
+.. _controller-entity-grid-create-edit-link:
+
+Linking the Data Grid
+---------------------
+
+Finally, you need to link both actions on the page that displays the list of tasks:
+
+**1. Add a link to create new tasks**
+
+The base ``index.html.twig`` template from the OroUIBundle that you already used to embed the data
```
Comment (yurio) [inline cookbook/entities/managing_tasks.rst]: Please specify that is the 'OroUIBundle:actions:index.html.twig' template

File: cookbook/entities/managing_tasks.rst [hunk 25]
```diff
@@ -0,0 +1,254 @@
...
+            'formErrors': form_errors(form) ? form_errors(form) : null,
+            'dataBlocks': dataBlocks,
+        } %}
+
+        {{ parent() }}
+    {% endblock content_data %}
+
+.. _controller-entity-grid-create-edit-link:
+
+Linking the Data Grid
```
Comment (yurio) [inline cookbook/entities/managing_tasks.rst]: I think you should change this title because create link dies not goes to the grid.

File: cookbook/entities/managing_tasks.rst (hunk 24 above)
Comment (yurio) [inline cookbook/entities/managing_tasks.rst]: Please add the link to the page where this index.html.twig template was used to better understanding.

File: cookbook/entities/managing_tasks.rst [hunk 26]
```diff
@@ -0,0 +1,254 @@
...
+grid comes with a pre-defined ``navButtons`` block which you can use to add a button that links to
+the *create task action*:
+
+.. code-block:: html+jinja
+    :linenos:
+
+    {# src/AppBundle/Resources/views/Task/index.html.twig #}
+    {% extends 'OroUIBundle:actions:index.html.twig' %}
+
+    {% set gridName = 'app-tasks-grid' %}
```
Comment (yurio) [inline cookbook/entities/managing_tasks.rst]: You does not use this variable in this case. Why did you add it?

File: cookbook/entities/managing_tasks.rst [hunk 27]
```diff
@@ -0,0 +1,254 @@
...
+    {% set pageTitle = 'Task' %}
```
Comment (yurio) [inline cookbook/entities/managing_tasks.rst]: You does not use this variable in this case. Why did you add it?

File: cookbook/entities/managing_tasks.rst [hunk 28]
```diff
@@ -0,0 +1,254 @@
...
+                id: ~
+                update_link:
+                    type: url
+                    route: app_task_update
+                    params:
+                        - id
+                # ...
+            actions:
+                # ...
+                delete:
```
Comment (yurio) [inline cookbook/entities/managing_tasks.rst]: You add an `update` link. Why the action was called `delete`?

File: cookbook/entities/deleting_tasks.rst (hunk 15 above)
Comment (yurio) [inline cookbook/entities/deleting_tasks.rst]: Please copy and modify the description from here https://github.com/orocrm/documentation/pull/206/files#diff-c9754199dc749289f99f93f7180443dcR229. It's much better than existing one.

File: cookbook/entities/deleting_tasks.rst [hunk 29]
```diff
@@ -0,0 +1,113 @@
...
+                    type:          delete
```
Comment (yurio) [inline cookbook/entities/deleting_tasks.rst]: Please add descriptions of this parameters

File: cookbook/entities/acl.rst [hunk 30]
```diff
@@ -0,0 +1,246 @@
+How to Protect Resources Using ACLs
```
Comment (yurio) [inline cookbook/entities/acl.rst]: At this file you describe how to protect only the entities. So i think, the `How to Protect Entities Using ACLs ` is much better.

**Comment by xabbuh** [comment]
@yurio The changes from the first commit were part of another PR (see #202) which was merged in the meantime. I took your comments into account and created #221 and #222 to address. Could you have a look at them?

File: cookbook/entities/acl.rst (hunk 8 above)
Comment (xabbuh) [inline cookbook/entities/acl.rst]: I think that is too much details for this cookbook. Anyone interested in learning more about the security concepts can refer to the book chapters and the configuration reference.

File: cookbook/entities/acl.rst (hunk 9 above)
Comment (xabbuh) [inline cookbook/entities/acl.rst]: I added a section that provides some more details: https://github.com/orocrm/documentation/pull/206/files#diff-f924240c2ac5eca3ac8df47bd7e79914R149

File: cookbook/entities/acl.rst [hunk 31]
```diff
@@ -0,0 +1,336 @@
...
+     *     }
+     * )
+     */
+    class Task
+    {
+        // ...
+    }
+
+After you have done this and have cleared the cache you can toggle all kinds of permission checks
+(``CREATE``, ``EDIT``, ``DELETE``, ``VIEW``, and ``ASSIGN`` in the user role management interface.
```
Comment (yurio) [inline cookbook/entities/acl.rst]: Please add ) after the list of permissions

File: cookbook/entities/acl.rst [hunk 32]
```diff
@@ -0,0 +1,336 @@
...
+
+        use Oro\Bundle\SecurityBundle\Annotation\Acl;
+        use Symfony\Bundle\FrameworkBundle\Controller\Controller;
+
+        class PageController extends Controller
+        {
+            /**
+             * @Acl(
+             *     id="app_static_pages",
+             *     type="action"
```
Comment (yurio) [inline cookbook/entities/acl.rst]: Please add parameter `label` and new parameter `group_name`

File: cookbook/entities/acl.rst [hunk 33]
```diff
@@ -0,0 +1,336 @@
...
+
+    When configuring the ACL using the YAML config format, you also have to set the controller to
+    bind the ACL to using the ``bindings`` option:
+
+    .. code-block:: yaml
+        :linenos:
+
+        # src/AppBunde/Resources/config/acl.yml
+        app_static_pages:
+            type: action
```
Comment (yurio) [inline cookbook/entities/acl.rst]: Please add parameter `label` and new parameter `group_name`

File: cookbook/entities/acl.rst [hunk 34]
```diff
@@ -0,0 +1,336 @@
...
+Once you have configured the ACLs you can protect all parts of your application. Anywhere in your
+PHP code you can use the ``isGranted()`` method of the ``oro_security.security_facade`` service
+(which is an instance of the :class:`Oro\\Bundle\\SecurityBundle\\SecurityFacade` class):
+
+.. code-block:: php
+    :linenos:
+
+    $securityFacade = $this->container->get('oro_security.security_facade');
+
+    if ($securityFacade->isGranted('app_task_create')) {
```
Comment (yurio) [inline cookbook/entities/acl.rst]: Please add more examples that will describe what you can set different types of data to this method

File: cookbook/entities/acl.rst [hunk 35]
```diff
@@ -0,0 +1,336 @@
...
+    :linenos:
+
+    {# src/AppBundle/Resources/views/Task/index.html.twig #}
+    {% extends 'OroUIBundle:actions:index.html.twig' %}
+
+    {% set gridName = 'app-tasks-grid' %}
+    {% set pageTitle = 'Task' %}
+
+    {% block navButtons %}
+        {% if resource_granted('app_task_create') %}
```
Comment (yurio) [inline cookbook/entities/acl.rst]: I think you can delete extends, set, block tags from this example. as the result, the example will be more readable.

And please add more examples with different data (entity, class name, etc)

File: cookbook/entities/acl.rst [hunk 36]
```diff
@@ -0,0 +1,336 @@
...
+
+    .. seealso::
+
+        All configuration options are explained in full details in the :doc:`/reference/annotation/acl`,
+        :doc:`/reference/annotation/acl_ancestor`, and :doc:`ACL YAML format </reference/format/acl>`
+        reference.
+
+.. _coobook-entities-acl-check:
+
+Perform Access Checks
```
Comment (yurio) [inline cookbook/entities/acl.rst]: Please add how to make protections with ACL Helper for ORM queries
